[CORE-12558] Auto Topic Sensor

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR introduces the `auto_topic_sensor`, a task that runs in a cluster link that will add mirror topics to the cluster link table.

The task will fetch a list of topics from the source cluster and then filter out topics based on filters supplied by the user and by authorization checks.  Once selected, the configurations for those topics are fetched from the source cluster and the topic and its properties are added to the cluster link table as a mirror topic.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
kafka: Added metadata_cache_info abstract class

Added an abstract class and metadata_cache_adapter that decouples
cluster::metadata_cache from the make_topic_configs function.  This
permits us to use that function in unit tests without having to create
an entire test fixture.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Added topic_configs to mirror topic metadata

Also added replication factor and partition count to mirror topic

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/model: Added configuration for auto topic task

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/model: Added utility to use filters

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Introduce source_topic_syncer

The purpose of this syncer task is to add mirror topics to the cluster
link table.  It will query for a list of topics from the source cluster
and then filter based on filters configured by the user and other
metadata.  If the topic is selected, it is added as a mirror topic to
the cluster link table.

Another task will run to reconcile these updates (to be implemented in a
future PR).

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/tests: Moved test_link_factory to deps

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/tests: Added function to await status report

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
c/cl: Updated error codes for frontend validation

Improved error code reporting coming from frontend validation

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
c/cl/tests: Improved base metadata creation

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Make get_link() protected

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Wire in topic_metadata_cache into link

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Add add_mirror_topic command to link_registry

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/server: Split out topic config utils

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/c/test: Added handling for describe_config in mock

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Fixed assertion when changing task interval

If the task interval changes while the task is running, this would have
rearmed the timer and then when the task finished, would attempt to
rearm an already armed timer, causing an assert.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/tests: Added more accessors to fixture

Also added ability to set topic configs on the test fixture

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/c/tests: Added ability to set controller_id

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/c: Added ability to track RF

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/c/tests: ALlow for zero replicas in cluster mock

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Add method to create mirror_topic to manager

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Wire in link_id and manager into link

Provide the ID of the link to the link class as well as a pointer back
to the owning manager.  This allows access to the link_registry, via the
manager, from the link so it can update the state of the cluster_link
when necessary.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented on src/v/cluster_link/task.cc:
```diff
@@ -55,8 +55,12 @@ class task::runner {
         // timepoint for when the timer should run
-        _timer.cancel();
-        _timer.arm((cur_timeout - _task->_run_interval) + interval);
+        if (_timer.armed()) {
```
wrong?

mmaslankaprv commented on src/v/cluster_link/task.cc:
```diff
@@ -55,8 +55,12 @@ class task::runner {
+            // will pick up the new interval once it completes
+            _timer.cancel();
+            _timer.arm((cur_timeout - _task->_run_interval) + interval);
+        }
```
nit: `timer.rearm`

mmaslankaprv commented on src/v/cluster_link/manager.cc:
```diff
@@ -233,6 +238,12 @@ ss::future<> manager::handle_on_leadership_change(
+ss::future<::cluster::cluster_link::errc> manager::add_mirror_topic(
+  model::id_t link_id, model::add_mirror_topic_cmd cmd) {
+    return _registry->add_mirror_topic(
+      link_id, std::move(cmd), ::model::timeout_clock::now() + 5s);
```
nit: add a define for timeout

mmaslankaprv commented on src/v/cluster_link/auto_topic_sensor.cc:
```diff
@@ -0,0 +1,337 @@
+          topic,
+          *partition_count,
+          *rf);
+
+        if (get_link()
+              ->get_topic_metadata_cache()
+              ->find_topic_cfg({::model::kafka_namespace, topic})
+              .has_value()) {
+            vlog(logger().trace, "Topic {} already exists", topic);
+            continue;
```
This is not directly related to this PR but i am wondering what will happen if this check succeeded but then somebody creates the topic which is already in the mirrored topics. I guess we will just stuck retrying the topic creation ?

michael-redpanda replied:
Good question.  What we're going to do for 25.3 is to disallow external topic creation if the topic name would be a match for the auto topic filters.  e.g. if we had a filter that says to include all topics prefixed with `foo-`, then any CreateTopics starting with `foo-` would be disallowed.

bharathv replied:
Also somewhat related question, since we are operating in the kafka API and the task runs periodically we cannot  distinguish between different revisions of same topic, right? 

<topic-partiitons=10> delete <topic-partitions=20> and the syncer ends up missing delete + recreate sequence, we may see an increase in partition count, guess thats classified as an edge case? (or maybe the deletes for "linked" topics are synchronous)?

michael-redpanda replied:
we will keep track of the topic ID... if the topic is deleted and re-created the topic ID will change.  It's not in here yet because RP doesn't support topic IDs, but will soon.

bharathv replied:
ah topic ids, that isn't registered in my head yet. ty.

michael-redpanda replied:
no worries - not out yet (but soon!!)

mmaslankaprv commented on src/v/cluster_link/auto_topic_sensor.h:
```diff
@@ -0,0 +1,64 @@
+#include "kafka/client/cluster.h"
+
+namespace cluster_link {
+class auto_topic_sensor : public task {
```
where this name comes from ? 
nit: would benefit from a short doc comment

bharathv replied:
also consider renames like topic_syncer or something along those lines.. (sensor Imo doesn't communicate the intent here).

michael-redpanda replied:
yeah happy to rename it - I picked 'sensor' because it's really just 'sensing' for new topics.

mmaslankaprv commented on src/v/cluster_link/auto_topic_sensor.cc:
```diff
@@ -0,0 +1,337 @@
+      "Fetching topic configs for {}",
+      std::views::keys(candidate_topics));
+
+    kafka::describe_configs_response response;
```
Since we are already describing config here is there a plan to add topic properties reconciliation here ?

michael-redpanda replied:
Yeah I'm still thinking about if I'm going to do that in this sensor or in a separate sensor task.

Applying the changes will be done in a reconciler outside of this.

bharathv commented:
+2.65k lines  🚢  🇮🇹

michael-redpanda commented:
> +2.65k lines 🚢 🇮🇹

I know I'm sorry

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--dd7648ac-c46c-4d46-a49f-929eb217ed8e-->
<!--TEST RESULTS-->
<!--9426734a-744a-4450-8648-ed259a9bdd20-->
<!--TEST RESULTS-->
<!--a3068c77-d821-4d21-af0e-7cca4827f9ea-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69902">build#69902</a></summary>


| test_class                                     | test_method                                    | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:-----------------------------------------------|:-----------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| src/v/kafka/client/test/cluster_test_with_mock | src/v/kafka/client/test/cluster_test_with_mock |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/69902#019858d1-a46b-4d5a-bfab-e0cb137f13ef | FAIL          | 0/1      |          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69983">build#69983</a></summary>


| test_class                     | test_method              | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------------|:-------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsMultiClusterTest | test_basic               | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/69983#01985dc2-8f07-4c19-991e-5e36d2530955 | FLAKY         | 20/21    | upstream reliability is '98.66369710467706'. current run reliability is '95.23809523809523'. drift is 3.4256 and the allowed drift is set to 50. The test should PASS  |
| DatalakeE2ETests               | test_json_schema_unicode | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"} | integration | https://buildkite.com/redpanda/redpanda/builds/69983#01985dbe-ecdd-44ff-93a2-36d94513d6ce | FLAKY         | 20/21    | upstream reliability is '99.29906542056075'. current run reliability is '95.23809523809523'. drift is 4.06097 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70167">build#70167</a></summary>


| test_class            | test_method                     | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------|:--------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest | test_higher_level_migration_api | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70167#019875a8-c300-4a48-9511-447514d871a2 | FLAKY         | 17/21    | upstream reliability is '98.97435897435898'. current run reliability is '80.95238095238095'. drift is 18.02198 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

bharathv commented:
> > +2.65k lines 🚢 🇮🇹
> 
> I know I'm sorry

all good, jk 😁

michael-redpanda commented:
> > > +2.65k lines 🚢 🇮🇹
> > 
> > 
> > I know I'm sorry
> 
> all good, jk 😁

No I honestly didn't realize how big it was until I pushed the PR, I thought it was like 1k...

mmaslankaprv commented:
@michael-redpanda i am wondering if the error is related with the changes in mock

michael-redpanda commented:
> @michael-redpanda i am wondering if the error is related with the changes in mock

yeah possibly - weird that it has only failed on the debug build and not release - i will look into it

michael-redpanda commented on src/v/cluster_link/auto_topic_sensor.cc:
```diff
@@ -0,0 +1,337 @@
+          "Failed to get supported API version for describe_configs: {}",
+          e.what());
+        vlog(logger().warn, "{}", msg);
+        std::ignore = change_state(model::task_state::link_unavailable, msg);
```
good catch Mr Ai bot

michael-redpanda committed:
k/c/tests: Added deferred stop to cluster

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* PR comments
* Fixed issue with cluster_test_with_mock

mmaslankaprv commented on src/v/kafka/client/test/cluster_mock.cc:
```diff
@@ -10,8 +10,110 @@
+class cluster_link_test_metadata_adapter : public kafka::metadata_cache_info {
+public:
+    ::model::compression get_default_compression() const override {
+        return config::shard_local_cfg().log_compression_type();
```
the tests should not query the config. Consider using a test config store or hardcoded values

michael-redpanda replied:
What I'm trying to do here is craft the response in the same way a describe configs response would look like in action.

mmaslankaprv replied:
this is fine, i just suggest not using the `configuration` to do it in tests

michael-redpanda replied:
oh gotcha, thanks for clarifying :)

mmaslankaprv commented on src/v/cluster_link/task.cc:
```diff
@@ -52,11 +52,15 @@ class task::runner {
+        if (_timer.armed()) {
+            // If the timer isn't armed, then we may be executing the task which
+            // will pick up the new interval once it completes
+            _timer.rearm(cur_timeout + (interval - previous_interval));
```
i agree with the bot, not sure how the timer behaves when it is passed with negative value

michael-redpanda replied:
Hilariously, this was the bot's suggestion that I applied with little sleep....

mmaslankaprv commented on src/v/cluster_link/auto_topic_sensor.cc:
```diff
@@ -0,0 +1,338 @@
+    } catch (const std::exception& e) {
+        auto msg = ssx::sformat("Failed to describe topics: {}", e.what());
+        vlog(logger().warn, "{}", msg);
+        std::ignore = change_state(model::task_state::link_unavailable, msg);
```
`std::move(msg)`

mmaslankaprv commented on src/v/cluster_link/auto_topic_sensor.cc:
```diff
@@ -0,0 +1,338 @@
+        configs.reserve(resp.configs.size());
+        for (const auto& c : resp.configs) {
+            if (c.value.has_value()) {
+                configs.emplace(c.name, *c.value);
```
move ?

mmaslankaprv commented on src/v/cluster_link/auto_topic_sensor.cc:
```diff
@@ -0,0 +1,338 @@
+          "Configs for topic {}: {}",
+          resp.resource_name,
+          configs);
+        auto it = candidate_topics.find(::model::topic{resp.resource_name});
```
consider creating local variable of type `model::topic` to prevent copying the string each time the `topic` is created

mmaslankaprv commented on src/v/cluster_link/auto_topic_sensor.cc:
```diff
@@ -0,0 +1,338 @@
+    chunked_hash_map<::model::topic, topic_metadata> candidate_topics;
+    candidate_topics.reserve(topics.size());
+
+    for (const auto& topic : topics) {
```
i am not a big fan of this loop as it makes multiple `find` calls into the topic cache. Maybe it would be better to simply expose the underlying map from the `topic_cache` and iterate over that instead ?

mmaslankaprv replied:
are you planning to address this comment ?

michael-redpanda replied:
I believe i have?

mmaslankaprv commented on src/v/cluster_link/model/types.cc:
```diff
@@ -157,14 +204,52 @@ auto fmt::formatter<
+    return fmt::format_to(ctx.out(), "{}", to_string_view(s));
+}
+
+auto fmt::formatter<cluster_link::model::topic_filter_pattern>::format(
```
I started thinking about the group migrations. Maybe we can name it `filter_pattern` so that it can be reused elsewhere ?

michael-redpanda replied:
fo sho

michael-redpanda replied:
`resource_name_filter_pattern`?

joe-redpanda commented on src/v/cluster/cluster_link/frontend.cc:
```diff
@@ -463,4 +482,78 @@ errc frontend::validator::validate_connection_config(
+          if (
+            p.pattern != ::cluster_link::model::topic_filter_pattern::wildcard
+            && !std::ranges::all_of(p.pattern, [](char c) {
+                   return std::isalnum(c) || c == '.' || c == '-' || c == '_';
```
Is this the utf check?
something like 'Öther' is valid utf8, but because 'Ö' is multibyte it won't pass this check

michael-redpanda replied:
Much more restrictive than utf-8.  Will be used against topic_names so it should be no more restrictive than the topic name check: https://github.com/redpanda-data/redpanda/blob/1eed6c7dd20fbbd1a6c40b67bf5d57d78d254fe5/src/v/model/validation.cc#L20-L38

bharathv commented on src/v/cluster_link/model/filter_utils.h:
```diff
@@ -0,0 +1,20 @@
+#include "cluster_link/model/types.h"
+
+namespace cluster_link::model {
+bool select_topic(
```
nit: method doc.

bharathv commented on src/v/cluster_link/tests/deps.h:
```diff
@@ -249,6 +249,13 @@ class cluster_link_manager_test_fixture {
     link_factory* get_link_factory() { return _lf; }
 
+    ss::future<std::optional<model::cluster_link_task_status_report>>
+    await_status_report(
```
nit: consider passing an abort source

bharathv commented on src/v/cluster_link/task.cc:
```diff
@@ -181,6 +185,11 @@ task::change_state(model::task_state new_state, ss::sstring reason) {
+        vlog(logger().trace, "Interval is unchanged, skipping update");
+        return;
+    }
+
```
> would attempt to rearm an already armed timer, causing an assert.

what assert is this?  this is the rearm logic.. AFAICT rearming an armed timer is totally ok

```
    void rearm(time_point until, std::optional<duration> period = {}) noexcept {
        if (_armed) {
            cancel();
        }
        arm(until, period);
    }
```

bharathv reviewed (commented):
lgtm, mostly nits

michael-redpanda commented on src/v/kafka/server/handlers/configs/config_utils.h:
```diff
@@ -40,27 +38,9 @@
-            ret.broker_changes.push_back(std::move(res));
-        };
-    }
-    return ret;
-}
+namespace kafka {
```
pre-existing code that was moved

michael-redpanda commented:
Force push (addressed PR comments):

* Renamed `topic_filter_pattern` to `resource_name_filter_pattern`
* Renamed `filters` field to `topic_name_filters`
* Renamed `auto_topic_sensor` to `source_topic_syncer` and renamed file
* Added comment blocks
* Addressed issue with time interval calculation
* Exposed the topic_data from the topic_cache

michael-redpanda commented:
Force push:

* Reworded topic syncer commit

michael-redpanda committed:
cl: Register source_topic_syncer_factory with manager

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Do not use global namespace configuration registry in cluster_mock - create and use its own
* Fixed a use after move error

michael-redpanda commented:
Force push:

* Fixed failing test

michael-redpanda commented:
Force push:

* Rebased off dev

michael-redpanda commented:
Force push:

* Fixed bad `BUILD` due to removal of `fragmented_vector`

michael-redpanda commented:
Force push:

* Fix up remaining issues with removal of fragmented_vector

mmaslankaprv commented on src/v/cluster_link/task.cc:
```diff
@@ -52,11 +52,21 @@ class task::runner {
+        // If the timer isn't armed, then we may be executing the task.  It
+        // will pick up the new interval once it finishes
+        if (_timer.armed()) {
+            auto new_timepoint = cur_timeout + (interval - previous_interval);
```
can we explicitly check if `interval > previous_interval` ? relying of overflow is not great, also it will trigger overflow sanitizer

michael-redpanda replied:
I think the problem here is not if `internval > previous_interval`, I think it's if `interval - previous_interval` results in a value `<= -cur_timeout` which would cause `new_timeout` to be negative, right?

mmaslankaprv replied:
yeah, you are right

michael-redpanda replied:
Addressed

michael-redpanda commented:
Force push:

* Improve timestamp calculation safety

mmaslankaprv commented on src/v/cluster_link/source_topic_syncer.cc:
```diff
@@ -0,0 +1,357 @@
+        co_return;
+    }
+
+    auto candidate_topics = find_candidate_topics();
```
Do we have plans on how to handle topics that were deleted from the source cluster ?
